[FLINK-3140] NULL value data layout in Row Serializer/Comparator

twalthr opened the pull request:
This PR implements FLINK-3140 and enables full NULL field support for grouping and sorting rows.

tillrohrmann commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/NullMaskUtils.scala:
```diff
@@ -0,0 +1,92 @@
+      }
+      // write byte
+      target.writeByte(b)
+    }
```
I think the same can be expressed a bit more sacalesque

```
value.productIterator.grouped(8).foreach{
      group =>
        val byte = group.foldLeft(0){
          (byteResult, element) =>
            (byteResult << 1) | (if (element == null) 0x01 else 0x00)
        }

        val finalResult = byte << (8 - group.length)

        target.writeByte(finalResult)
    }
```

tillrohrmann commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/NullMaskUtils.scala:
```diff
@@ -0,0 +1,92 @@
+    }
+  }
+
+  def readIntoAndCopyNullMask(len: Int, source: DataInputView, target: DataOutputView,
```
Line breaking of parameter lists in Scala

tillrohrmann commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/RowComparator.scala:
```diff
@@ -0,0 +1,468 @@
+    /** serializers to deserialize the first n fields for comparison */
+    val serializers: Array[TypeSerializer[_]],
+    /** auxiliary fields for normalized key support */
+    private var normalizedKeyLengths: Array[Int],
```
Why is this a `var`? Can't it be a `val`?

tillrohrmann commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/RowComparator.scala:
```diff
@@ -0,0 +1,468 @@
+      serializers: Array[TypeSerializer[_]]) = {
+    this(keyPositions, comparators, serializers, new Array[Int](keyPositions.length), 0, 0, false)
+    // set up auxiliary fields for normalized key support
+    setupAuxiliaryFields()
```
Can't we simply let `setupAuxiliaryFields` return the `normalizedKeyLengths`, `numLeadingNormalizableKeys` and `normalizableKeyPrefixLen` values for the constructor call? Maybe we should then rename the method to `createAuxiliaryFields`.

tillrohrmann commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/RowComparator.scala:
```diff
@@ -0,0 +1,468 @@
+
+  // cache for the deserialized key field objects
+  @transient
+  private var deserializedKeyFields1: Array[Any] = null
```
We could make this an `Option` to get rid of the `null` value.

twalthr replied:
Can I really use Option here? What happens after deserialization? What is the default value of a transient field in Scala?

tillrohrmann replied:
Hmm that's a legitimate objection. What about using a lazy initializer and a `val` instead? Then we would also get rid of the `var`.

tillrohrmann commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/RowComparator.scala:
```diff
@@ -0,0 +1,468 @@
+      }
+      i += 1
+    }
+  }
```
We could do it like this

```
def createAuxiliaryFields(
     comparators: Array[TypeComparator[_]])
    : (List[Int], Int, Int) = {
    comparators.takeWhile {
      c =>
        c.supportsNormalizedKey() && c.invertNormalizedKey() == comparators(0).invertNormalizedKey()
    }.foldLeft((List[Int](), 0, 0)) {
      case ((normKeyLengths, numLeadingNormKeys, normKeyPrefixLen), comparator) =>
        val len = comparator.getNormalizeKeyLen

        if (len < 0) {
          throw new RuntimeException("Comparator " + comparator.getClass.getName +
            " specifies an invalid length for the normalized key: " + len)
        }

        ((len + 1) :: normKeyLengths, numLeadingNormKeys + 1, normKeyPrefixLen + len + 1)
    }
  }
```

twalthr replied:
Yes you are right. I just tried to be as close as possible to TupleComparatorBase to prevent bugs. IMHO I think my implementation is more readable as it similar to other classes (e.g. https://github.com/dataArtisans/cascading-flink/blob/master/src/main/java/com/dataartisans/flink/cascading/types/tuple/DefinedTupleComparator.java). But if other people think that your solution is better, I will change it.

tillrohrmann replied:
I would assume that Scala programmer will say the opposite since it is easier to reason about this method (no side effects whatsoever).

twalthr commented:
Thanks for your feedback @tillrohrmann.
Many parts could be expressed in more Scala style, but for performance reasons and to prevent bugs in this crucial component I chose simple while loops and tried to be as close as possible to already proven code (esp. TupleSerializerBase and Fabians NullMaskSerDeUtils). In FLINK-2203 @aljoscha already mentioned that: "Could you please change this back to the simple while loop. I know that using the fancy Scala features is tempting but the performance of a simple while loop should be better than creating a range and iterating over it using foreach."

tillrohrmann commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/RowComparator.scala:
```diff
@@ -0,0 +1,468 @@
+      case iobex: IndexOutOfBoundsException =>
+        throw new KeyFieldOutOfBoundsException(keyPositions(i))
+    }
+  }
```
We could write this as

```
override def compareToReference(referencedComparator: TypeComparator[Row]): Int = {
    val other: RowComparator = referencedComparator.asInstanceOf[RowComparator]

    comparators.zip(other.comparators).view.map{
      case (comparator, otherComparator) =>
        val nullCheck = comparator.asInstanceOf[TypeComparator[Any]].equalToReference(nullChecker)
        val nullCheckOther = otherComparator.asInstanceOf[TypeComparator[Any]]
          .equalToReference(nullChecker)

        (nullCheck, nullCheckOther) match {
          case (true, true) => 0
          case (true, false) =>
            comparator match {
              case basicComp: BasicTypeComparator[_] if basicComp.isAscendingComparison =>
                1
              case basicComp: BasicTypeComparator[_] if !basicComp.isAscendingComparison =>
                -1
              case _ => 1
            }
          case (false, true) =>
            comparator match {
              case basicComp: BasicTypeComparator[_] if basicComp.isAscendingComparison =>
                -1
              case basicComp: BasicTypeComparator[_] if !basicComp.isAscendingComparison =>
                1
              case _ => -1
            }
          case (false, false) =>
            comparator.compareToReference(otherComparator)
        }
    }.dropWhile(_ == 0).headOption.getOrElse(0)
  }
```

twalthr replied:
Are you sure that your code is more perfomant than mine? E.g. "(nullCheck, nullCheckOther) match" isn't that creating an objects internally?

tillrohrmann replied:
Yeah I just thought about it myself. I guess with the zipping and so on, this is not the most efficient implementation. Therefore, I've also deleted this comment here.

tillrohrmann commented:
@aljoscha is absolutely right about the performance aspect. However, the key difference is that he was referring in his comment to the implementation of the `serialize` method which is extensively used and, thus, should be as efficient as possible.

If I'm not mistaken, then this argument does not apply to the `createAuxiliaryFields` method, though. The reason is that this method is called before the job is actually executed and, consequently, should not have any impact on the runtime behaviour. But I might be wrong here, since we're also shipping `TypeInformation` to the cluster, nowadays.

tillrohrmann commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/NullMaskUtils.scala:
```diff
@@ -0,0 +1,92 @@
+      while (bytePos < 8) {
+        b = b << 1
+        bytePos += 1
+      }
```
If you're really concerned about performance, then this should be `b <<= (8 - bytePos)`.

tillrohrmann commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/NullMaskUtils.scala:
```diff
@@ -0,0 +1,92 @@
+      b = 0x00
+      // set bits in byte
+      bytePos = 0
+      while (bytePos < 8 && fieldPos < len) {
```
Furthermore, this should then be `while (bytePos < numPos)` with `int numPos = Math.min(8, len - fieldPos)`. The `fieldPos` update should then happen outside of the loop `fieldPos += numPos`. That way you save 7 comparisons, 87 additions and 7 assignments.

twalthr commented:
I implemented your feedback. But I don't know if we should use `Option` for the `deserializedKeyFields`, because we then need to add checks for each serialized comparison. What do you think?

twalthr commented:
I now use lazy vals for that.

fhueske commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/RowSerializer.scala:
```diff
@@ -40,15 +46,32 @@ class RowSerializer(val fieldSerializers: Array[TypeSerializer[Any]])
+    }
+
+    if (from.productArity != len || reuse.productArity != len) {
       throw new RuntimeException("Row arity of reuse and from do not match.")
```
Error message should be: "Row arity of reuse or from is incompatible with this RowSerializer".

fhueske commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/RowComparator.scala:
```diff
@@ -0,0 +1,463 @@
+
+      // deserialize field 1
+      if (nullMask1(i)) {
+        deserializedKeyFields1(i) = null
```
If you use `nullMask1` during the comparison step to check if a field `null` is, you don't have to set the field to `null` here and read the object the next time without a `reuse` object.

fhueske commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/RowComparator.scala:
```diff
@@ -0,0 +1,463 @@
+
+      var cmp = 0
+      // both values are null -> equality
+      if (firstElement == null && secondElement == null) {
```
use `nullMask1` and `nullMask2` here instead of checking the field value

fhueske commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/RowComparator.scala:
```diff
@@ -0,0 +1,463 @@
+            " specifies an invalid length for the normalized key: " + len)
+        }
+
+        val newNormKeyLengths = (len + 1) :: normKeyLengths // add one for a null byte
```
`len` can be `Integer.MAX_VALUE` (as for example for `StringComparator`). Adding 1 results in an overflow.

fhueske commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/RowComparator.scala:
```diff
@@ -0,0 +1,463 @@
+
+        val newNormKeyLengths = (len + 1) :: normKeyLengths // add one for a null byte
+        val newNumLeadingNormKeys = numLeadingNormKeys + 1
+        val newNormKeyPrefixLen = normKeyPrefixLen + len + 1 // add one for a null byte
```
`newNormKeyPrefix` can become negative if one comparator returns `Integer.MAX_VALUE` for `getNormalizeKeyLen`.

fhueske commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/RowComparator.scala:
```diff
@@ -0,0 +1,463 @@
+}
+
+object RowComparator {
+  private class NullChecker extends Comparable[AnyRef] with Serializable {
```
I don't think it is a good idea to use the `NullChecker` for reference comparisons against the field comparators. Doesn't this rely on a specific behavior of the field comparators? The field comparator might call the `equals` method on the null-valued reference object instead of the other object or it may throw a NPE if null is set as reference object.
 
I think there are two better alternatives:

1. Track which fields of the reference object are null using a Boolean array within the RowComparator and don't set null values in the field comparator.
2. Wrap each field comparator with a null-aware comparator. This comparator could also take the logic for the null handling of normalized keys.

fhueske commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/RowComparator.scala:
```diff
@@ -0,0 +1,463 @@
+        // one value is null -> inequality
+        // order is considered for basic types
+        else if (nullCheck || nullCheckOther) {
+          if (comparator.isInstanceOf[BasicTypeComparator[_]]) {
```
You can access the sort orders of the individual fields by overwriting the `createComparator` method in the `RowTypeInfo` and directly passing this information to the `RowComparator`.
Then you can avoid the `isInstanceOf` check for basic types and have a generic solution.

fhueske commented on flink-core/src/main/java/org/apache/flink/api/common/typeutils/base/BasicTypeComparator.java:
```diff
@@ -42,6 +42,10 @@ protected BasicTypeComparator(boolean ascending) {
 		this.ascendingComparison = ascending;
 	}
 
+	public boolean isAscendingComparison() {
```
There is no need to touch this class, if you overwrite the `createComparator` method in the `RowTypeInfo` (details below).

fhueske commented on flink-staging/flink-table/src/test/scala/org/apache/flink/api/table/typeinfo/RowSerializerTest.scala:
```diff
@@ -0,0 +1,125 @@
+  class RowSerializerTestInstance(
+      serializer: TypeSerializer[Row],
+      testData: Array[Row])
+    extends SerializerTestInstance[Row](serializer, classOf[Row], -1, testData: _*) {
```
It would be good to have a little bit more test data than 3 rows.

twalthr commented:
Thanks for your feedback @fhueske. I will rework this PR during this week.

twalthr commented:
@fhueske I updated the PR again. I implemented your suggestion of NullAwareComparators wrapping the normal comparators. I also added additional test cases. What do you think?

fhueske commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/NullAwareComparator.scala:
```diff
@@ -0,0 +1,186 @@
+  override def supportsSerializationWithKeyNormalization(): Boolean = false
+
+  override def writeWithKeyNormalization(record: T, target: DataOutputView): Unit =
+    throw new UnsupportedOperationException("Normalized keys not supported.")
```
Exception should be changed to something like "Record serialization with leading normalized keys not supported"

fhueske commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/NullAwareComparator.scala:
```diff
@@ -0,0 +1,186 @@
+    throw new UnsupportedOperationException("Normalized keys not supported.")
+
+  override def readWithKeyDenormalization(reuse: T, source: DataInputView): T =
+    throw new UnsupportedOperationException("Normalized keys not supported.")
```
Exception should be changed to something like "Record deserialization with leading normalized keys not supported"

twalthr replied:
Does it make sense to change it in the other Comparators as well? `TupleComparatorBase` doesn't even have a exception message.

fhueske replied:
I would not do it in this PR. It is already complex enough ;-) 
IMO, it is not a big issue and super important to be fixed. The method would only be called if the system ignores the result of `supportsSerializationWithKeyNormalization()`.

I just pointed it out here, because it is part of the PR.

fhueske commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/NullAwareComparator.scala:
```diff
@@ -0,0 +1,186 @@
+    throw new UnsupportedOperationException("Normalized keys not supported.")
+
+  override def isNormalizedKeyPrefixOnly(keyBytes: Int): Boolean =
+    wrappedComparator.isNormalizedKeyPrefixOnly(keyBytes)
```
should be `keyBytes - 1`.

fhueske commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/NullAwareComparator.scala:
```diff
@@ -0,0 +1,186 @@
+ * Null-aware comparator that wraps a comparator which does not support null references.
+ */
+class NullAwareComparator[T](
+    val wrappedComparator: TypeComparator[T],
```
We must ensure that the `wrappedComparator` is for atomic types only. 
Wrapping will not work for composite types!

fhueske replied:
I think we can check for atomic types at this place, but should at least do it when a `NullAwareComparator` is created.

twalthr replied:
Are you sure we want to limit rows to atomic types only?
Why should the wrapping not work for composite types? If you have a look into my test cases, I already test Tuples and Pojos successfully.

fhueske replied:
I think right now, the `extractKeys` and `getFlatComparators()` methods do not work as expected. These methods are used to compare records from different types, i.e., in a Join. For the comparison, the key fields are extracted by `extractKeys` method that corresponds with the input and then compared using the flat comparators. Hence, the number of keys and comparators must always be the same and the comparators must be able to handle null values. If we want to support composite types as keys, we must handle the case of null-valued fields very carefully, i.e., return the right number of null values for each nested field + all nested comparators must support null values.

fhueske replied:
Since all comparators returned by `getFlatComparators()` must be atomic, we could wrap these again in a `NullAwareComparator`.

fhueske commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/NullAwareComparator.scala:
```diff
@@ -0,0 +1,186 @@
+
+  // ----------------------------------------------------------------------------------------------
+
+  override def getFlatComparators: Array[TypeComparator[_]] =
```
This method must be implemented!
It should be implemented as `return new TypeComparator[] {this};`, i.e., return itself. This is OK, because the wrapped comparator is atomic and therefore already flat.

fhueske replied:
OK, I see. Your implementation assumes that this class is only used within `RowComparator`. How about guarding against other uses and making this a private member class of `RowComparator`?

twalthr replied:
The method is not implemented because the row comparator will take care of it.

twalthr replied:
Sorry, I haven't seen your comment. Ok, I will make the class private.

fhueske commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/NullAwareComparator.scala:
```diff
@@ -0,0 +1,186 @@
+    throw new UnsupportedOperationException("Comparator does not support extraction of flat " +
+      "comparators.")
+
+  override def compareSerialized(firstSource: DataInputView, secondSource: DataInputView): Int =
```
This method must be implemented. It is called in the `NormalizedKeySorter` for instance.

fhueske commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/NullAwareComparator.scala:
```diff
@@ -0,0 +1,186 @@
+  override def extractKeys(record: Any, target: Array[AnyRef], index: Int): Int = {
+    if (record == null) {
+      target(index) = null
+      1
```
This does only work, if the wrapped comparator is atomic. Otherwise, we will get a different number of keys for `null` and `non-null` composite types. Maybe, we can create a default instance using the type serializer and pass it to the comparator to create the right number of values and later overwrite all with `null`. This is hacky, but might work :-/.

fhueske commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/RowComparator.scala:
```diff
@@ -0,0 +1,422 @@
+  // --------------------------------------------------------------------------------------------
+
+  override def getFlatComparator(flatComparators: util.List[TypeComparator[_]]): Unit =
+    comparators.map(_.wrappedComparator).foreach {
```
The key fields extracted by `extractKeys()` can be null. These keys are compared by the comparators returned by this method. Hence, we must return comparators which are null-aware.

fhueske commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/RowComparator.scala:
```diff
@@ -0,0 +1,422 @@
+        i += 1
+      }
+    } catch {
+      case npex: NullPointerException =>
```
This exception should be removed. If we encounter a NPE here, it is not because the key field was null but because the system was not able to handle the null correctly.

fhueske commented on flink-staging/flink-table/src/test/scala/org/apache/flink/api/table/typeinfo/RowComparatorTest.scala:
```diff
@@ -0,0 +1,112 @@
+import org.apache.flink.api.table.typeinfo.RowComparatorTest.MyPojo
+import org.junit.Assert._
+
+class RowComparatorTest extends ComparatorTestBase[Row] {
```
`extractKeys` and `getFlatComparators` are not tested by the `ComparatorTestBase`. It would be good to add (combined) tests for these methods with `null`-valued atomic and composite fields.

fhueske commented:
Hi Timo, the PR looks good except for the `extractKeys` and `getFlatComparator` methods. I think it is possible to support composite type wrapping with the `NullAwareComparator` but we need to take care of composite types in `extractKeys` and `getFlatComparators`.

fhueske commented on flink-staging/flink-table/src/test/scala/org/apache/flink/api/table/typeinfo/RowComparatorTest.scala:
```diff
@@ -0,0 +1,112 @@
+      BasicTypeInfo.INT_TYPE_INFO,
+      BasicTypeInfo.DOUBLE_TYPE_INFO,
+      BasicTypeInfo.STRING_TYPE_INFO,
+      new TupleTypeInfo[tuple.Tuple1[Int]](BasicTypeInfo.INT_TYPE_INFO),
```
Can we use a composite type with more than one internal field, for example a `Tuple2`?

twalthr commented:
@fhueske Thanks for reviewing my code so intensively. Your suggestion sounds very good. But again I have the problem of how to access the order of the atomic types after flattening. The `NullAwareComparator` needs this information. Would it be so bad to add `isAscending()` to basic types or do you have a better idea?

fhueske commented:
The `NullAwareComparator` knows about the order of its wrapped comparator. So we can implement the `getFlatComparators()` method of `NullAwareComparator`, call `getFlatComparators()` on the wrapped comparator and wrap each returned comparator again in a `NullAwareComparator` with the same order as the wrapping one. I think that should work, but maybe I am overlooking something here...

IMO, the problem with `isAscending()` is that it is only implemented for BasicTypes (Int, String, etc.) but not for other atomic types such as generic or writable types.

twalthr commented:
Sorry, I mixed up atomic and basic types. I forgot that a flat comparators list could also contain GenericTypes etc.

Can we assume that the orders of the flat comparators returned by the `wrappedComparator` are equal to the orders of the `wrappedComparator` itself?

fhueske commented:
Yes, I think so. 
ATM, the Table API does not allow to access nested fields but only Row fields as a whole. So there is no way to define different order for nested types. I think we'll keep it like this for now. If we change it at some point, we have to touch the code again.

twalthr commented:
@fhueske I have implemented your feedback and rebased it onto the current master. Unfortunately, Git had problems with the new package structure. I had to squash my commits. You can find the most important changes in the extractKeys() methods und the RowComparatorTest. I added a test to ComparatorTestBase and extended my tests for RowComparator/RowSerializer..

fhueske commented:
No problem, @twalthr. Thanks for the update! Will review it in the next days.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/NullAwareComparator.scala:
```diff
@@ -0,0 +1,231 @@
+  }
+
+  /**
+   * This method does not support records that are null. Use [[extractKeysFromNull()]] instead.
```
We can get the number of fields to extract by calling `getFlatComparators()` in the constructor of the `NullAwareComparator` and use this number to fill the target array with nulls if `record` is `null`. 
Then we do not need the special `extractKeysFromNull` method and avoid some overhead.

twalthr replied:
Sometimes I think to complicated ;-)

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/RowComparator.scala:
```diff
@@ -0,0 +1,423 @@
+
+  override def duplicate(): TypeComparator[Row] = {
+    // copy comparator and serializer factories
+    val comparatorsCopy = comparators.map(_.duplicate().asInstanceOf[NullAwareComparator[_]])
```
Isn't this wrapping `NullAwareComparator`s in `NullAwareComparator`s?

twalthr replied:
No there is no wrapping. The `duplicate()` methods creates a duplicate without wrapping.

fhueske replied:
Ah, yes. You're right.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/RowComparator.scala:
```diff
@@ -0,0 +1,423 @@
+      while(i < keyPositions.length) {
+        code *= TupleComparatorBase.HASH_SALT(i & 0x1F)
+        val element = value.productElement(keyPositions(i)) // element can be null
+        code += comparators(i).asInstanceOf[TypeComparator[Any]].hash(element)
```
I think you can you get rid of these casts if you define `comparators` as `Array[NullAwareComparator[Any]]` instead of `Array[NullAwareComparator[_]]`.

fhueske replied:
If it works, this should also apply for `serializers`.

twalthr replied:
Unfortunetaly I could not get rid of all casts. but I tried to simplify it a bit.

fhueske commented:
PR looks good. Just a few minor things to fix before it can be merged.

fhueske commented:
Looks good to merge :-)

twalthr commented:
I will merge it, once the build succeeded.
